core/src/Cuda/Kokkos_CudaSpace.cpp

```diff
@@ -209,19 +214,28 @@ void *CudaSpace::impl_allocate(
     const char *arg_label, const size_t arg_alloc_size,
     const size_t arg_logical_size,
     const Kokkos::Tools::SpaceHandle arg_handle) const {
+  return impl_allocate(Kokkos::Cuda{}, arg_label, arg_alloc_size,
+                       arg_logical_size, arg_handle);
+}
+void *CudaSpace::impl_allocate(
+    const Cuda &exec_space, const char *arg_label, const size_t arg_alloc_size,
+    const size_t arg_logical_size,
+    const Kokkos::Tools::SpaceHandle arg_handle) const {
   void *ptr = nullptr;
 
 #ifndef CUDART_VERSION
 #error CUDART_VERSION undefined!
 #elif (defined(KOKKOS_ENABLE_IMPL_CUDA_MALLOC_ASYNC) && CUDART_VERSION >= 11020)
   cudaError_t error_code;
   if (arg_alloc_size >= memory_threshold_g) {
-    error_code = cudaMallocAsync(&ptr, arg_alloc_size, 0);
-    KOKKOS_IMPL_CUDA_SAFE_CALL(cudaDeviceSynchronize());
+    cudaStream_t stream = exec_space.cuda_stream();
+    error_code          = cudaMallocAsync(&ptr, arg_alloc_size, stream);
+    KOKKOS_IMPL_CUDA_SAFE_CALL(cudaStreamSynchronize(stream));
```

Comment:
we can't just do this: this is a change in behavior which we need to discuss more. Specifically the base allocate will now not fence everything. I think this would be more agreeable if the version without executions space stays with the old behavior.

Reply:
Sure, we should discuss that but we still fence the stream/execution space provided anyway.